File: codeforces/1600~1699/1684/c.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

void solve() {
  int n = 0, m = 0;
  cin >> n >> m;
  vector<vector<int>> a(n, vector<int>(m));
  for (int i = 0; i < n; ++i) {
    for (int j = 0; j < m; ++j) {
      cin >> a[i][j];
    }
  }

  set<pair<int,int>> s;
  for (int i = 0; i < n; ++i) {
    vector<int> row = a[i];
    sort(row.begin(), row.end());
    int diff = 0;
    vector<int> index;
    for (int j = 0; j < m; ++j) {
      if (row[j] != a[i][j]) {
        diff++;
        index.push_back(j);
      }
    }

    if (diff != 2 && diff != 0) {
      cout << "-1" << endl;
      return;
    }

    if (diff == 2) {
      s.insert({index[0]+1, index[1]+1});
    }
  }

  if (s.empty()) {
    cout << "1 1" << endl;
  } else if (s.size() > 1) {
    cout << "-1" << endl;
  } else {
    pair<int,int> index = *s.begin();

    for (int i = 0; i < n; ++i) {
      swap(a[i][index.first-1], a[i][index.second-1]);
      for (int j = 1; j < m; ++j) {
        if (a[i][j-1] > a[i][j]) {
          cout << "-1" << endl;
          return;
        }
      }
    }

    cout << index.first << " " << index.second << endl;
  }
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  int T = 0;
  cin >> T;
  for (int t = 0; t < T; ++t) {
    solve();
  }
  return 0;
}

File: codeforces/1600~1699/1684/c.rs
use std::io;
use std::str::FromStr;
use std::fmt::Debug;
use std::collections::BTreeSet;

fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    io::stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    io::stdin().read_line(&mut s).unwrap();
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

fn solve() {
    let inputs = read_vec::<i32>();
    let n = inputs[0] as usize;
    let m = inputs[1] as usize;

    let mut a: Vec<Vec<i32>> = Vec::new();
    for _i in 0..n {
        let row = read_vec::<i32>();
        a.push(row);
    }


    let mut s: BTreeSet<(usize,usize)> = BTreeSet::new();
    for i in 0..n {
        let mut row = a[i].clone();
        let mut index: Vec<usize> = Vec::new();
        row.sort();

        let mut diff = 0;
        for j in 0..m {
            if row[j] != a[i][j] {
                diff += 1;
                index.push(j);
            }
        }

        if diff != 2 && diff != 0 {
            println!("-1");
            return;
        }

        if diff == 2 {
            s.insert((index[0]+1, index[1]+1));
        }
    }

    if s.is_empty() {
        println!("1 1");
    } else if s.len() > 1 {
        println!("-1");
    } else {
        let index = s.pop_first().unwrap();
        let i1 = index.0 - 1;
        let i2 = index.1 - 1;
        for i in 0..n {
            a[i].swap(i1, i2);
            for j in 1..m {
                if a[i][j-1] > a[i][j] {
                    println!("-1");
                    return;
                }
            }
        }

        println!("{} {}", index.0, index.1);
    }
}

fn main() {
    let n = read_int::<i32>();
    for _t in 0..n {
        solve();
    }
}
